File: Codeforces/Codeforces Round #490 (Div. 3)/D - Equalize the Remainders.py
n, m = map(int, input().split())
a = list(map(int, input().split()))

mod = [[] for i in range(m)]
for i in range(n):
    mod[a[i] % m].append(i)
#print(mod)
k = n // m
free = []
for i in range(m):
    if len(mod[i]) > k:
        free.extend(mod[i][k:])
        mod[i] = mod[i][:k]
ans = 0
#print(free)
#print(mod)
for i in range(m):
    mlen = len(mod[i])
    if mlen < k:
        for j in free[:k - mlen]:
            ans += (i - a[j]) % m
            a[j] += (i - a[j]) % m
            mod[i].append(j)
        free = free[k - mlen:]
#print(mod)
print(ans)
print(' '.join([str(i) for i in a]))

File: Codeforces/Codeforces Round #490 (Div. 3)/D - Equalize the Remainders.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
    int n, m;
    cin >> n >> m;
    vector<int> a(n);
    vector<vector<int>> mod(m, vector<int>());
    for (int i = 0; i < n; ++i) {
        cin >> a[i];
        mod[a[i] % m].push_back(i);
    }
    int k = n / m;
    vector<int> free;
    for (int i = 0; i < m; ++i) {
        if (mod[i].size() > k) {
            while (mod[i].size() > k) {
                free.push_back(mod[i].back());
                mod[i].pop_back();
            }
        }
    }
    int ans = 0;
    for (int i = m - 2, j = 0; j < m; i = (i - 1 + m) % m, j++) {
        if (mod[i].size() < k) {
            while (mod[i].size() < k) {
                int tmp = ((i - a[free.back()]) % m + m) % m;
                ans += tmp;
                a[free.back()] +=  tmp;
                mod[i].push_back(free.back());
                free.pop_back();
            }
        }
    }
    cout << ans << '\n';
    for (auto e : a) {
        cout << e << ' ';
    }
    return 0;
}